[JSC] Fix behavior of `Intl.DurationFormat` for negative values

sosukesuzuki opened the pull request:
#### 3f74bba5c41c78732611942b04ba1d43a5fa030c
<pre>
[JSC] Fix behavior of `Intl.DurationFormat` for negative values
<a href="https://bugs.webkit.org/show_bug.cgi?id=275095">https://bugs.webkit.org/show_bug.cgi?id=275095</a>

Reviewed by Ross Kirsling.

This patch includes the following two related fixes:

1. When the values of multiple units are negative, display `-` only for the largest unit.
For example, the duration `{ years: -1, months: -2 }` should be formatted as `&quot;-1 years, 2 months&quot;`.
According to the spec[1], this behavior is controlled by the `signDisplayed` variable (initially
set to true). This variable is set to false once any unit is formatted.

2. Implement the DurationSign abstract operation [2] to display negative zero.
For example, the duration `{ hours: 0, seconds: -1 }` should be formatted as `&quot;-0 hours, 1 second&quot;`.
This is described in section 4.f.iii.3.b of the spec [1] as follows:

    &gt; b. If value is 0 and DurationSign(duration.[[Years]], duration.[[Months]], duration.[[Weeks]], duration.[[Days]], duration.[[Hours]], duration.[[Minutes]], duration.[[Seconds]], duration.[[Milliseconds]], duration.[[Microseconds]], duration.[[Nanoseconds]]) is -1, then
    &gt;     i. Set value to negative-zero.

The `partitionDurationFormatPattern` function [3] in the `harness/testintl.js` of test262 describes
a similar process to the changes in this patch.

[1]: <a href="https://tc39.es/proposal-intl-duration-format/#sec-partitiondurationformatpattern">https://tc39.es/proposal-intl-duration-format/#sec-partitiondurationformatpattern</a>
[2]: <a href="https://tc39.es/proposal-intl-duration-format/#sec-durationsign">https://tc39.es/proposal-intl-duration-format/#sec-durationsign</a>
[3]: <a href="https://github.com/tc39/test262/blob/249657722525cc8e43b1ddb91f8df0b4b011fcf6/harness/testIntl.js#L2649-L2662">https://github.com/tc39/test262/blob/249657722525cc8e43b1ddb91f8df0b4b011fcf6/harness/testIntl.js#L2649-L2662</a>

* JSTests/stress/intl-duration-format-negative-values.js: Added.
(sameValue):
(throw.new.Error):
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/IntlDurationFormat.cpp:
(JSC::getDurationSign):
(JSC::collectElements):

Canonical link: <a href="https://commits.webkit.org/279902@main">https://commits.webkit.org/279902@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b9bb2a6a1587cb16c68ab88bac9568978a316940

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54894 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34339 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/7480 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58172 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5625 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/41887 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/5649 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/44458 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3812 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56989 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/32443 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47559 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25585 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/29233 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/4900 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3766 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/48261 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/51080 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5120 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59762 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/54399 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/30151 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/5269 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/51880 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/31288 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47637 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/51299 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/text/state-changed (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12057 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/32303 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/66695 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31076 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12719 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

rkirsling commented on Source/JavaScriptCore/runtime/IntlDurationFormat.cpp:
```diff
@@ -322,6 +322,26 @@ struct Element {
+{
+    for (unsigned index = 0; index < numberOfTemporalUnits; ++index) {
+        TemporalUnit unit = static_cast<TemporalUnit>(index);
+        double value = duration[unit];
```
You don't actually need to think about specific `TemporalUnit`s here. 😄 
```suggestion
    for (auto value : duration) {
```

rkirsling commented on Source/JavaScriptCore/runtime/IntlDurationFormat.cpp:
```diff
@@ -330,8 +350,10 @@ static Vector<Element> collectElements(JSGlobalObject* globalObject, const IntlD
+    bool signDisplayed = true;
     String separator;
     Vector<Element> elements;
+    std::optional<DurationSignType> durationSign = std::nullopt;
```
`std::nullopt` is the default value for `std::optional`, so no need to write it explicitly.
```suggestion
    std::optional<DurationSignType> durationSign;
```

rkirsling commented on Source/JavaScriptCore/runtime/IntlDurationFormat.cpp:
```diff
@@ -406,6 +428,18 @@ static Vector<Element> collectElements(JSGlobalObject* globalObject, const IntlD
+            if (!signDisplayed)
+                skeletonBuilder.append(" +_"_s);
+            else if (!value) {
+                if (durationSign == std::nullopt)
```
```suggestion
                if (!durationSign)
```

rkirsling commented on Source/JavaScriptCore/runtime/IntlDurationFormat.cpp:
```diff
@@ -406,6 +426,18 @@ static Vector<Element> collectElements(JSGlobalObject* globalObject, const IntlD
+            else if (!value) {
+                if (!durationSign)
+                    durationSign = getDurationSign(duration);
+                if (durationSign == DurationSignType::Negative) {
```
Interesting; I would have thought that this would be constant `false` because the LHS and RHS types are different, but apparently it works. I still feel like it would be clearer to write `durationSign.value()`, but I don't think we have a style rule about this, so maybe it's okay.

rkirsling commented on Source/JavaScriptCore/runtime/IntlDurationFormat.cpp:
```diff
@@ -434,6 +466,7 @@ static Vector<Element> collectElements(JSGlobalObject* globalObject, const IntlD
                 return formattedNumber;
             };
 
+
```
Oops, let's remove this extra line too.

rkirsling commented on Source/JavaScriptCore/runtime/IntlDurationFormat.cpp:
```diff
@@ -330,8 +348,10 @@ static Vector<Element> collectElements(JSGlobalObject* globalObject, const IntlD
     auto scope = DECLARE_THROW_SCOPE(vm);
 
     bool done = false;
+    bool signDisplayed = true;
```
It seems confusing to start out by saying that we already _did_ something; I think this would be clearer as `bool needsSignDisplay = false;`

sosukesuzuki replied:
I used the variable names as they are in the specification, but I think your suggestion is clearer and better. Thank you.

rkirsling reviewed (changes requested):
Sorry, I have one more round of feedback. 🙈

sosukesuzuki committed:
[JSC] Fix behavior of `Intl.DurationFormat` for negative values
https://bugs.webkit.org/show_bug.cgi?id=275095

Reviewed by Ross Kirsling.

This patch includes the following two related fixes:

1. When the values of multiple units are negative, display `-` only for the largest unit.
For example, the duration `{ years: -1, months: -2 }` should be formatted as `"-1 years, 2 months"`.
According to the spec[1], this behavior is controlled by the `signDisplayed` variable (initially
set to true). This variable is set to false once any unit is formatted.

2. Implement the DurationSign abstract operation [2] to display negative zero.
For example, the duration `{ hours: 0, seconds: -1 }` should be formatted as `"-0 hours, 1 second"`.
This is described in section 4.f.iii.3.b of the spec [1] as follows:

    > b. If value is 0 and DurationSign(duration.[[Years]], duration.[[Months]], duration.[[Weeks]], duration.[[Days]], duration.[[Hours]], duration.[[Minutes]], duration.[[Seconds]], duration.[[Milliseconds]], duration.[[Microseconds]], duration.[[Nanoseconds]]) is -1, then
    >     i. Set value to negative-zero.

The `partitionDurationFormatPattern` function [3] in the `harness/testintl.js` of test262 describes
a similar process to the changes in this patch.

[1]: https://tc39.es/proposal-intl-duration-format/#sec-partitiondurationformatpattern
[2]: https://tc39.es/proposal-intl-duration-format/#sec-durationsign
[3]: https://github.com/tc39/test262/blob/249657722525cc8e43b1ddb91f8df0b4b011fcf6/harness/testIntl.js#L2649-L2662

* JSTests/stress/intl-duration-format-negative-values.js: Added.
(sameValue):
(throw.new.Error):
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/IntlDurationFormat.cpp:
(JSC::getDurationSign):
(JSC::collectElements):

Canonical link: https://commits.webkit.org/279902@main
